api/pkg/s3/objectrestore.go

```diff
@@ -0,0 +1,129 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"encoding/xml"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/opensds/multi-cloud/s3/pkg/model"
+
+	"github.com/emicklei/go-restful"
+	"github.com/micro/go-log"
+	. "github.com/opensds/multi-cloud/s3/pkg/exception"
+	s3 "github.com/opensds/multi-cloud/s3/proto"
+	"golang.org/x/net/context"
+
+	"github.com/opensds/multi-cloud/api/pkg/policy"
+	. "github.com/opensds/multi-cloud/api/pkg/utils/constants"
+)
+
+func (s *APIService) ObjectRestore(request *restful.Request, response *restful.Response) {
+	if !policy.Authorize(request, response, "object:restore") {
+		return
+	}
+	url := request.Request.URL
+	log.Logf("URL is %v", request.Request.URL.String())
+
+	/*
+		parsing the parameters from the URL request
+	*/
+	bucketName := request.PathParameter("bucketName")
+	objectKey := request.PathParameter("objectKey")
+
+	if strings.HasSuffix(url.String(), "/") {
+		objectKey = objectKey + "/"
+	}
+
+	/*
+		parsing the request body from the query
+	*/
+	body := ReadBody(request)
+	if body != nil {
+
+		/*
+			unmarshal restore object xml struct into s3 struct
+		*/
+		createRestoreConf := model.RestoreRequest{}
+		err := xml.Unmarshal(body, &createRestoreConf)
+		if err != nil {
+			response.WriteError(http.StatusInternalServerError, err)
+			return
+		}
+		ctx := context.Background()
+		/*
+			validate if specified bucket exists or not
+		*/
+		BucketMD, _ := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
+		if BucketMD == nil {
+			response.WriteError(http.StatusInternalServerError, fmt.Errorf("error: specified bucket does not exist"))
+			return
+		}
+		/*
+			creating Get object input
+		*/
+		objectInput := s3.GetObjectInput{Bucket: bucketName, Key: objectKey}
+
+		/*
+			getting the object metadata from database using GET object input
+		*/
+		objectMD, _ := s.s3Client.GetObject(ctx, &objectInput)
+
+		if objectMD != nil {
+			client := getBackendByName(s, objectMD.Backend)
+			/*
+				creating the restore object input
+			*/
+			restoreInput := s3.RestoreObjectInput{Key: objectKey, Bucket: bucketName, Days: createRestoreConf.Days, Tier: createRestoreConf.GlacierJobParameters.Tier}
+
+			/*
+				call for restoring the object
+			*/
+			s3restore := s3.ObjectRestoreStatus{}
+			s3restore.RestoreMarker = true
+
+			s3err := client.RestoreObject(&restoreInput, ctx)
+			if s3err.Code != ERR_OK {
+				s3restore.RestoreState = ActionStateRestoreFailed
+				response.WriteError(http.StatusInternalServerError, s3err.Error())
+				return
+			} else {
+				s3restore.RestoreState = ActionStateRestoring
+			}
+
+			objectMD.RestoreStatus = &s3restore
+
+			res, err := s.s3Client.UpdateObject(ctx, objectMD)
```

Comment:
In case the system crashed before calling cloud API to restore object but before updating metadata in database, then the restoring status maybe lost. So we need to update metadata in database before calling cloud API to restore an object.